resource/ui/ts/util/querycache.ts

```diff
@@ -0,0 +1,130 @@
+         * given query function. It is expected that the query function returns
+         * a promise for results.
+         */
+        constructor(private _query:() => promise<T>) {
+            this.refresh()
+        }
+            
+        /**
+         * Refresh invokes the underlying query function, unless another
+         * invocation is already in progress. The currently cached results (if
+         * any) are not replaced until the query invocation completes.
+         */
+        refresh() {
+            // Clear outstanding result if it has already returned.
+            this.processOutstanding();
+            if (!this._outstanding) {
+                this._outstanding = {
+                    result:this._query(),
+                    error:m.prop(<Error> null), 
+                }
+                this._outstanding.result.then(null, this._outstanding.error);
+            }
+        }
+
+        /**
+         * setQuery replaces the query function underlying this QueryCache.
+         */
+        setQuery(q:() => promise<T>) {
+            this._query = q;
+        }
+
+        /**
+         * hasData returns true if at least one query has completed. This
+         * indicates that either the result() or error() functions will return
+         * non-null.
+         */
+        hasData():boolean {
+            this.processOutstanding();
+            return this._epoch > 0;
+        }
+
+        /**
+         * result returns the result of most recent invocation of the underlying
+         * query. If the most recent invocation returned an error, this method will
+         * return null.
+         */
+        result():T {
+            this.processOutstanding();
+            return this._result;
+        }
+
+        /**
+         * result returns any error that resulted from the most recent
+         * invocation of the underlying query. If the most recent invocation did
+         * not return an error, this method will return null.
+         */
+        error():Error {
+            this.processOutstanding();
+            return this._error;
+        }
+
+        /**
+         * epoch returns the current epoch of the cached query. The epoch is
+         * incremented each time the cached result is refreshed; the first
+         * result has an epoch of 1.
+         */
+        epoch():number {
+            this.processOutstanding();
+            return this._epoch;
+        }
+
+        /** 
+         * This structure will be non-null when a query is in-flight, or has 
+         * completed but not been processed. When an in-flight query
+         * completes, one and only one of these fields will contain a value.
+         */
+        private _outstanding:{
+            result:promise<T>;
+            error:property<Error>;
+        }
+
+        /**
+         * Check for a completed outstanding query, replacing cached results
+         * with the result.
+         */
+        private processOutstanding() {
```

Comment:
Why do you call this with every function instead of chaining it to when the query completes?